tools/Google.Cloud.SampleUtil/ApplicationChooser.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.SampleUtil
+{
+    /// <summary>
+    /// Class allowing a user to choose a sample to run, after
+    /// examining an assembly for all classes containing a static Main method, either
+    /// parameterless or with a string array parameter.
+    /// The ApplicationChooser is designed to run in both interactive mode (not 
+    /// implemented yet) and non-interactive mode.
+    /// In the interactive mode, the options are presented in a list ordered 
+    /// first by those with a Description attribute (by that description) and 
+    /// then those without (ordered by name).
+    /// In the non-interactive mode, the ApplicationChooser will invoke the sample
+    /// with the sample file name (the first positional commandline argument).
+    /// </summary>
+    public class ApplicationChooser
+    {
+        /// <summary>
+        /// Looks up a sample based on its name, and invoke it with the remaining arguments.
+        /// </summary>	
+        public static void Run(string[] args)
+        {
+            if (args.Length == 0)
+            {
+                throw new NotImplementedException("Interative mode isn't implemented yet.");
+            }
+            string sample = args[0];
+            var assembly = Assembly.GetEntryAssembly();
+
+            // Create a dictionary, so that we can easily look up a sample entry point based on
+            // the filename of the sample.
+            var samples = assembly
+                          .DefinedTypes
+                          .Select(t => GetEntryPoint(t))
+                          .Where(ep => ep != null && ep != assembly.EntryPoint)
+                          .ToDictionary(ep => ep.DeclaringType.Name.Replace("Main", string.Empty), ep => ep);
```

Comment:
I think it might be better to write a method that trims "Main" from the end if it exists - otherwise this could cause problems if we ever have samples to do with "Maintenance" for example.

Reply:
Good point!